content_sync/apis/github.py

```diff
@@ -177,42 +173,19 @@ def upsert_content_files_for_user(  # pylint:disable=too-many-locals
             if filepath:
                 synced_filepaths.update({sync_state.id: filepath})
                 data = self.format_content_to_file(content)
-                modified_element_list.append(
-                    InputGitTreeElement(filepath, "100644", "blob", data)
+                modified_element_list.extend(
+                    self.get_tree_elements(sync_state, data, filepath)
                 )
-                # Remove the old filepath stored in the sync state data if it doesn't match current path
-                if (
-                    sync_state.data
-                    and sync_state.data.get(GIT_DATA_FILEPATH, None)
-                    and sync_state.data[GIT_DATA_FILEPATH] != filepath
-                ):
-                    modified_element_list.append(
-                        InputGitTreeElement(
-                            sync_state.data[GIT_DATA_FILEPATH],
-                            "100644",
-                            "blob",
-                            sha=None,
-                        )
-                    )
 
         if len(modified_element_list) > 0:
-            tree = repo.create_git_tree(modified_element_list, base_tree)
-            parent = repo.get_git_commit(main_sha)
-            git_user = self.git_user(User.objects.filter(id=user_id).first())
-            commit = repo.create_git_commit(
-                "Sync all content",
-                tree,
-                [parent],
-                committer=git_user,
-                author=git_user,
-                **kwargs,
+            commit = self.commit_tree(
+                modified_element_list, User.objects.filter(id=user_id).first()
             )
-            main_ref.edit(commit.sha)
 
             # Mark all as synced
             for synced_state_id, filepath in synced_filepaths.items():
                 sync_state = ContentSyncState.objects.get(id=synced_state_id)
-                sync_state.current_checksum = content.calculate_checksum()
+                sync_state.current_checksum = sync_state.content.calculate_checksum()
```

Comment:
Actually, this now introduces a race condition where the content could change between when we read it for the git tree and when we read it here. You could probably add it to the data `synced_filepaths`.

Reply:
Good point, how about something like this?
```python
                synced_filepaths.update(
                    {
                        sync_state.id: {
                            "filepath": filepath, 
                            "checksum": content.calculate_checksum()
                        }
                    }
                )
``` 

and then:

```python
            for synced_state_id, sync_info in synced_filepaths.items():
                sync_state = ContentSyncState.objects.get(id=synced_state_id)
                sync_state.data = {GIT_DATA_FILEPATH: sync_info["filepath"]}
                if sync_state.content.calculate_checksum() == sync_info["checksum"]:
                    sync_state.current_checksum = sync_info["checksum"]
                    sync_state.mark_synced()
                else:
                    sync_state.save()
```

Reply:
I think we need to move away from `mark_synced()`, in fact remove it, cause I don't think there's any benefit and it can in fact cause issues. I think instead you want this to be:


```python
for synced_state_id, sync_info in synced_filepaths.items():
                sync_state = ContentSyncState.objects.get(id=synced_state_id)
                sync_state.data = {GIT_DATA_FILEPATH: sync_info["filepath"]}
                sync_state.synced_checksum = sync_info["checksum"]
                sync_state.save()
```